### DISCUSSION TITLE
Is it possible to define a class with arbitrary extra keys, or alternatively a Mapping with some required keys?

### DISCUSSION THREAD

**Discussion by edgarrmondragon** [apple/pkl#661] [Q&A] [answered]
I'm trying to build a template for [Python logging configuration](https://docs.python.org/3/library/logging.config.html#logging-config-dict-userdef).

The problem:

Each of the values in the `handlers` mapping has a required `class` key, a few optional pre-defined keys, and zero or more arbitrary keys which are passed through as keyword arguments to the handler’s constructor defined in `class`.

I'm stuck defining the Pkl class for that

```pkl
class Handler {
    `class`: String
    level: String?
    formatter: String?
    filters: List<String>?
    // TODO: Support extra handler keyword arguments
}
```

So I'm currently forced to use a `Mapping`, which loses a lot of the information compared to the class:

```pkl
handlers = new Mapping {
    ["stdout"] = new Mapping {
        ["class"] = "logging.StreamHandler"
        ["level"] = "INFO"
        ["formatter"] = "simple"
        ["stream"] = "ext://sys.stdout"
    }
    ["stderr"] = new Mapping {
        ["class"] = "logging.StreamHandler"
        ["level"] = "ERROR"
        ["formatter"] = "simple"
        ["stream"] = "ext://sys.stderr"
    }
    ["file"] = new Mapping {
        ["class"] = "logging.FileHandler"
        ["formatter"] = "simple"
        ["filename"] = "app.log"
        ["mode"] = "w"
    }
}
```

So, it is possible to define a class with arbitrary extra keys, or alternatively a Mapping with some required keys?

Thanks!

**Comment by HT154** [comment, marked as answer]
One way to do this is to have a `Mapping` property in your class for those extra fields and combine everything using an output converter, eg.

```pkl
class Handler {
    `class`: String
    level: String?
    formatter: String?
    filters: List<String>?
    kwargs: Mapping<String(this != "class", this != "level", this != "formatter", this != "filters"), Any> // constrain keys to prevent overlap with predefined properties
}

output {
  renderer {
    converters {
      [Handler] = (it) -> it.toMap().remove("kwargs") + it.kwargs.toMap() // merge types properties and kwargs
    }
  }
}
```

**Reply by edgarrmondragon** [reply]
Clean and easy. Thanks!
